InterfacePython/TasmanianSG.py

```diff
@@ -2365,3 +2362,159 @@ def copyGrid(source, iOutputsBegin = 0, iOutputsEnd = -1):
+            False: the read failed, check the CLI output for an error message
+        '''
+        effective_filename = bytes(sFilename, encoding='utf8') if sys.version_info.major == 3 else sFilename
+        if self.pLibTSG.tsgReadCustomTabulated(self.pCustomTabulated, c_char_p(effective_filename)) == 0:
+            raise TasmanianInputError("sFilename", "ERROR: {0:1s} does not appear to be a valid Tasmanian file.".format(sFilename))
+
+    def write(self, sFilename, bUseBinaryFormat = True):
+        '''
+        Writes the CustomTabulated object to a file.
+
+        sFilename: string indicating a location where the CustomTabulated instance will be written to.
+        bUseBinaryFormat: boolean
+            True: write to a binary file
+            False: write to an ASCII file
+        '''
+        effective_filename = bytes(sFilename, encoding='utf8') if sys.version_info.major == 3 else sFilename
+        self.pLibTSG.tsgWriteCustomTabulated(self.pCustomTabulated, c_char_p(effective_filename), c_int(bUseBinaryFormat))
+
+    def getNumLevels(self):
+        return self.pLibTSG.tsgGetNumLevelsCustomTabulated(self.pCustomTabulated)
+
+    def getNumPoints(self, level):
+        return self.pLibTSG.tsgGetNumPointsCustomTabulated(self.pCustomTabulated, c_int(level))
+
+    def getIExact(self, level):
+        return self.pLibTSG.tsgGetIExactCustomTabulated(self.pCustomTabulated, c_int(level))
+
+    def getQExact(self, level):
+        return self.pLibTSG.tsgGetQExactCustomTabulated(self.pCustomTabulated, c_int(level))
+
+    def getDescription(self):
+        return self.pLibTSG.tsgGetDescriptionCustomTabulated(self.pCustomTabulated)
+
+    def getWeightsNodes(self, level):
+        '''
+        Outputs the weights and nodes (in that order) of the CustomTabulated instance for a given level.
+
+        level: int indicating the level where the weights/nodes will be pulled from.
+
+        output: int (weights), int (nodes)
+        '''
+        iNumPoints = self.getNumPoints(level)
+        if (iNumPoints == 0):
+            return np.empty([0], np.float64), np.empty([0], np.float64)
+        weights, nodes = np.zeros([iNumPoints]), np.zeros([iNumPoints])
+        self.pLibTSG.tsgGetWeightsNodesStaticCustomTabulated(self.pCustomTabulated, c_int(level), np_arr_to_ctype(weights, np.float64),
+                                                             np_arr_to_ctype(nodes, np.float64))
+        return weights, nodes
+
+
+def makeCustomTabulatedFromFile(filename, tasmanian_library=0):
+    '''
+    Wrapper that calls the CustomTabulated constructor which reads its data from a file.
+
+    filename: string indicating the location of the file.
+    tasmanian_library: see getTsgLib().
+
+    output: CustomTabulated
+    '''
+    ct = CustomTabulated(tasmanian_library)
+    ct.read(filename)
+    return ct
+
+def makeCustomTabulatedFromData(num_levels, num_nodes, precision, nodes, weights, description, tasmanian_library=0):
```

Comment:
This isn't directly related to the exotic quadrature, but it may be nice to have. It's OK to add.

Reply:
Noted.